File: src/winnow/desktop.py
"""Register a .desktop launcher + icons so the file manager offers
'Open With -> Winnow' for folders and images.

Works from any install method because the icons ship inside the package
(``winnow.resources``); the launcher points at whichever ``winnow`` is running.
"""

from __future__ import annotations

import os
import shutil
import subprocess
import sys
from importlib.resources import files
from pathlib import Path

ICON_SIZES = (48, 64, 128, 256)
_MIME = ";".join(
    [
        "inode/directory",
        "image/jpeg",
        "image/png",
        "image/bmp",
        "image/gif",
        "image/tiff",
        "image/webp",
    ]
)


def _winnow_exe() -> str:
    """Absolute path to the currently running winnow entry point."""
    if getattr(sys, "frozen", False):  # PyInstaller binary
        return os.path.abspath(sys.executable)
    found = shutil.which("winnow")
    if found:
        return os.path.abspath(found)
    return os.path.abspath(sys.argv[0] or "winnow")


def install_desktop() -> Path:
    """Write the launcher + icons into ~/.local and refresh caches. Returns the
    path to the installed .desktop file."""
    apps = Path.home() / ".local/share/applications"
    icons = Path.home() / ".local/share/icons/hicolor"
    apps.mkdir(parents=True, exist_ok=True)

    res = files("winnow.resources")
    for sz in ICON_SIZES:
        data = (res / f"winnow-{sz}.png").read_bytes()
        dest = icons / f"{sz}x{sz}" / "apps"
        dest.mkdir(parents=True, exist_ok=True)
        (dest / "winnow.png").write_bytes(data)

    exe = _winnow_exe()
    desktop = apps / "winnow.desktop"
    desktop.write_text(
        "[Desktop Entry]\n"
        "Type=Application\n"
        "Version=1.0\n"
        "Name=Winnow\n"
        "GenericName=Image Culling Tool\n"
        "Comment=Fast keyboard-driven image culling / selection\n"
        f"Exec={exe} %f\n"
        "Icon=winnow\n"
        "Terminal=false\n"
        "Categories=Graphics;Viewer;2DGraphics;\n"
        f"MimeType={_MIME};\n"
    )
    desktop.chmod(0o755)

    for cmd in (
        ["update-desktop-database", str(apps)],
        ["gtk-update-icon-cache", "-f", str(icons)],
    ):
        try:
            subprocess.run(cmd, check=False, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        except FileNotFoundError:
            pass
    return desktop

File: crates/winnow-gui/src/app/desktop.rs
//! `winnow --install-desktop`: register the "Open With -> Winnow" launcher and
//! icons so folders and images get an entry in the file manager. Icons are
//! embedded in the binary, so this works from any install location.

use std::io;
use std::path::PathBuf;
use std::process::Command;

const ICONS: &[(u32, &[u8])] = &[
    (48, include_bytes!("../../icons/winnow-48.png")),
    (64, include_bytes!("../../icons/winnow-64.png")),
    (128, include_bytes!("../../icons/winnow-128.png")),
    (256, include_bytes!("../../icons/winnow-256.png")),
];

const MIME: &str = "inode/directory;image/jpeg;image/png;image/bmp;image/gif;image/tiff;image/webp;";

fn exe_path() -> String {
    std::env::current_exe()
        .ok()
        .and_then(|p| p.canonicalize().ok())
        .map(|p| p.to_string_lossy().into_owned())
        .unwrap_or_else(|| "winnow".to_string())
}

pub fn install_desktop() -> io::Result<PathBuf> {
    let home = PathBuf::from(std::env::var("HOME").map_err(|_| {
        io::Error::new(io::ErrorKind::NotFound, "HOME not set")
    })?);
    let apps = home.join(".local/share/applications");
    let icons = home.join(".local/share/icons/hicolor");
    std::fs::create_dir_all(&apps)?;

    for (sz, data) in ICONS {
        let dir = icons.join(format!("{sz}x{sz}")).join("apps");
        std::fs::create_dir_all(&dir)?;
        std::fs::write(dir.join("winnow.png"), data)?;
    }

    let desktop = apps.join("winnow.desktop");
    let contents = format!(
        "[Desktop Entry]\n\
         Type=Application\n\
         Version=1.0\n\
         Name=Winnow\n\
         GenericName=Image Culling Tool\n\
         Comment=Fast keyboard-driven image culling / selection\n\
         Exec={exe} %f\n\
         Icon=winnow\n\
         Terminal=false\n\
         Categories=Graphics;Viewer;2DGraphics;\n\
         MimeType={mime}\n",
        exe = exe_path(),
        mime = MIME,
    );
    std::fs::write(&desktop, contents)?;

    // Refresh caches (best effort; absent on some desktops).
    let _ = Command::new("update-desktop-database").arg(&apps).status();
    let _ = Command::new("gtk-update-icon-cache").arg("-f").arg(&icons).status();

    Ok(desktop)
}
